pandas/src/interval.pyx

```diff
@@ -0,0 +1,131 @@
+cimport numpy as np
+import numpy as np
+import pandas as pd
+
+cimport cython
+import cython
+
+from cpython.object cimport (Py_EQ, Py_NE, Py_GT, Py_LT, Py_GE, Py_LE,
+                             PyObject_RichCompare)
+
+
+_VALID_CLOSED = frozenset(['left', 'right', 'both', 'neither'])
+
+
+cdef class IntervalMixin:
+    property closed_left:
+        def __get__(self):
+            return self.closed == 'left' or self.closed == 'both'
+
+    property closed_right:
+        def __get__(self):
+            return self.closed == 'right' or self.closed == 'both'
+
+    property open_left:
+        def __get__(self):
+            return not self.closed_left
+
+    property open_right:
+        def __get__(self):
+            return not self.closed_right
+
+    property mid:
+        def __get__(self):
+            try:
+                return 0.5 * (self.left + self.right)
+            except TypeError:
+                # datetime safe version
+                return self.left + 0.5 * (self.right - self.left)
+
+
+cdef _interval_like(other):
+    return (hasattr(other, 'left')
+            and hasattr(other, 'right')
+            and hasattr(other, 'closed'))
+
+
+cdef class Interval(IntervalMixin):
```

Comment:
Maybe more appropriate for a future PR.

Is there any interest in adding an intersection method for this class?
For instance
```python
from pandas.core.interval import Interval
region1 = Interval(0, 100)
region2 = Interval(50, 200)
overlap = region1.intersect(region2)
assert overlap == Interval(50, 100)
```